examples/games/stepping.rs

```diff
@@ -0,0 +1,343 @@
+use bevy::app::MainScheduleOrder;
+use bevy::ecs::schedule::{NodeId, ScheduleLabel};
+use bevy::prelude::*;
+use std::collections::HashMap;
+
+/// Plugin to add a stepping UI to an example
+#[derive(Default)]
+pub struct SteppingPlugin {
+    schedule_labels: Vec<Box<dyn ScheduleLabel>>,
+    top: Val,
+    left: Val,
+}
+
+impl SteppingPlugin {
+    /// Initialize the plugin to step the schedules specified in `labels`
+    pub fn for_schedules(labels: Vec<Box<dyn ScheduleLabel>>) -> SteppingPlugin {
+        SteppingPlugin {
+            schedule_labels: labels,
+            ..default()
+        }
+    }
+
+    /// Set the location of the stepping UI when activated
+    pub fn at(self, left: Val, top: Val) -> SteppingPlugin {
+        SteppingPlugin { top, left, ..self }
+    }
+}
+
+impl Plugin for SteppingPlugin {
+    fn build(&self, app: &mut App) {
+        // create and insert our stepping schedule into the main schedule order
+        app.init_schedule(Stepping);
+        let mut order = app.world.resource_mut::<MainScheduleOrder>();
+        order.insert_after(Update, Stepping);
+
+        // add our startup & stepping systems
+        app.insert_resource(State {
+            schedule_labels: self.schedule_labels.clone(),
+            ui_top: self.top,
+            ui_left: self.left,
+            status: Status::Init,
+            system_text_map: HashMap::new(),
+            last_system_ids: Vec::new(),
+        })
+        .add_systems(
+            Stepping,
+            (
+                build_ui.run_if(not(initialized)),
+                handle_input.run_if(initialized),
+                update_ui.run_if(initialized),
+            )
+                .ignore_stepping(),
+        );
+    }
+}
+
+/// Independent [`Schedule`] for stepping systems.
+///
+/// The stepping systems must run in their own schedule to be able to inspect
+/// all the other schedules in the [`App`].  This is because the currently
+/// executing schedule is removed from the [`Schedules`] resource while it is
+/// being run.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, ScheduleLabel)]
+struct Stepping;
+
+#[derive(Debug)]
+enum Status {
+    Init,
+    Run,
+    Step(usize),
+}
+
+/// Struct for maintaining stepping state
+#[derive(Resource, Debug)]
+struct State {
+    // schedules that will be stepped
+    schedule_labels: Vec<Box<dyn ScheduleLabel>>,
+
+    // map of Schedule/NodeId to TextSection index in UI entity
+    // This is used to draw the position indicator as we step
+    system_text_map: HashMap<(Box<dyn ScheduleLabel>, NodeId), usize>,
+
+    // keep track of the last system NodeId in each schedule; we use this to
+    // know when to switch to the next schedule when stepping
+    last_system_ids: Vec<NodeId>,
+
+    // status of the stepping plugin
+    status: Status,
+
+    // ui positioning
+    ui_top: Val,
+    ui_left: Val,
+}
```

Comment:
@alice-i-cecile Is it an anti-pattern to merge this state with `SteppingPlugin`, and using that as a Resource?

Reply:
Yeah, don't pun by making things both plugins and resources. That tends to confuse users who don't have a great grasp on traits very fast.